### PR TITLE
Notary review

### PR DESCRIPTION
**Pull request by mtrmac**

### CODE REVIEW

File: notary/notary.spec [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+BuildRequires: golang(github.com/Sirupsen/logrus)
+BuildRequires: golang(github.com/Sirupsen/logrus/hooks/bugsnag)
+BuildRequires: golang(github.com/agl/ed25519)
+BuildRequires: golang(github.com/bugsnag/bugsnag-go)
+BuildRequires: golang(github.com/docker/distribution/context)
+BuildRequires: golang(github.com/docker/distribution/health)
+BuildRequires: golang(github.com/docker/distribution/registry/api/errcode)
+BuildRequires: golang(github.com/docker/distribution/registry/api/v2)
+BuildRequires: golang(github.com/docker/distribution/registry/auth)
+BuildRequires: golang(github.com/docker/distribution/registry/auth/silly)
```
Comment (ingvagabund) [inline notary/notary.spec]: golang(github.com/docker/distribution/registry/auth/silly) can be removed, it is required by tests only

File: notary/notary.spec [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+BuildRequires: golang(github.com/go-sql-driver/mysql)
+BuildRequires: golang(github.com/golang/protobuf/proto)
+BuildRequires: golang(github.com/google/gofuzz)
+BuildRequires: golang(github.com/gorilla/mux)
+BuildRequires: golang(github.com/jinzhu/gorm)
+BuildRequires: golang(github.com/mattn/go-sqlite3)
+BuildRequires: golang(github.com/miekg/pkcs11)
+BuildRequires: golang(github.com/prometheus/client_golang/prometheus)
+BuildRequires: golang(github.com/spf13/viper)
+BuildRequires: golang(github.com/stretchr/testify/assert)
```
Comment (ingvagabund) [inline notary/notary.spec]: golang(github.com/stretchr/testify/assert) can be removed, it is required by tests only

File: notary/notary.spec [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+BuildRequires: golang(github.com/stretchr/testify/require)
```
Comment (ingvagabund) [inline notary/notary.spec]: golang(github.com/stretchr/testify/require) can be removed, it is required by tests only

File: notary/notary.spec [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+Requires:      golang(github.com/Sirupsen/logrus)
+Requires:      golang(github.com/Sirupsen/logrus/hooks/bugsnag)
+Requires:      golang(github.com/agl/ed25519)
+Requires:      golang(github.com/bugsnag/bugsnag-go)
+Requires:      golang(github.com/docker/distribution/context)
+Requires:      golang(github.com/docker/distribution/health)
+Requires:      golang(github.com/docker/distribution/registry/api/errcode)
+Requires:      golang(github.com/docker/distribution/registry/api/v2)
+Requires:      golang(github.com/docker/distribution/registry/auth)
+Requires:      golang(github.com/docker/distribution/registry/auth/silly)
```
Comment (ingvagabund) [inline notary/notary.spec]: Can be removed

File: notary/notary.spec [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+Requires:      golang(github.com/go-sql-driver/mysql)
+Requires:      golang(github.com/golang/protobuf/proto)
+Requires:      golang(github.com/google/gofuzz)
+Requires:      golang(github.com/gorilla/mux)
+Requires:      golang(github.com/jinzhu/gorm)
+Requires:      golang(github.com/mattn/go-sqlite3)
+Requires:      golang(github.com/miekg/pkcs11)
+Requires:      golang(github.com/prometheus/client_golang/prometheus)
+Requires:      golang(github.com/spf13/viper)
+Requires:      golang(github.com/stretchr/testify/assert)
```
Comment (ingvagabund) [inline notary/notary.spec]: Can be removed

File: notary/notary.spec [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+Requires:      golang(github.com/stretchr/testify/require)
```
Comment (ingvagabund) [inline notary/notary.spec]: Can be removed

File: notary/notary.spec [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+Provides:      golang(%{import_path}/signer/client) = %{version}-%{release}
+Provides:      golang(%{import_path}/signer/keydbstore) = %{version}-%{release}
+Provides:      golang(%{import_path}/signer/keys) = %{version}-%{release}
+Provides:      golang(%{import_path}/trustmanager) = %{version}-%{release}
+Provides:      golang(%{import_path}/trustmanager/yubikey) = %{version}-%{release}
+Provides:      golang(%{import_path}/tuf) = %{version}-%{release}
+Provides:      golang(%{import_path}/tuf/client) = %{version}-%{release}
+Provides:      golang(%{import_path}/tuf/data) = %{version}-%{release}
+Provides:      golang(%{import_path}/tuf/encrypted) = %{version}-%{release}
+Provides:      golang(%{import_path}/tuf/keys) = %{version}-%{release}
```
Comment (ingvagabund) [inline notary/notary.spec]: This is invalid provides. tuf directory does not defines keys package

File: notary/notary.spec [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+BuildArch:       noarch
+
+%if 0%{?with_check}
+#Here comes all BuildRequires: PACKAGE the unit tests
+#in %%check section need for running
+%endif
+
+# test subpackage tests code from devel subpackage
+Requires:        %{name}-devel = %{version}-%{release}
+
```
Comment (ingvagabund) [inline notary/notary.spec:199]: Please append

%if 0%{?with_check} && ! 0%{?with_bundled}
BuildRequires: golang(github.com/docker/distribution/registry/auth/silly)
BuildRequires: golang(github.com/spf13/cobra)
BuildRequires: golang(github.com/stretchr/testify/assert)
BuildRequires: golang(github.com/stretchr/testify/require)
%endif

Requires:      golang(github.com/docker/distribution/registry/auth/silly)
Requires:      golang(github.com/spf13/cobra)
Requires:      golang(github.com/stretchr/testify/assert)
Requires:      golang(github.com/stretchr/testify/require)

File: notary/notary.spec [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+#define license tag if not already defined
+%{!?_licensedir:%global license %doc}
+
+%if 0%{?with_devel}
+%files devel -f devel.file-list
+%license LICENSE
+%dir %{gopath}/src/%{provider}.%{provider_tld}/%{project}
+# This is not caught by the generated devel.file-list
+%dir %{gopath}/src/%{import_path}/cmd
```
Comment (ingvagabund) [inline notary/notary.spec]: cmd directory does not contain any file with 'package cmd' directive. All files has 'package main'. For that, none of the files can be imported. This, cmd does not define importable golang package.

Please remove.

Comment (ingvagabund) [inline notary/notary.spec]: Looking into those files, all of them are used only for building cmd/main.go.

File: notary/notary.spec (hunk 1 above)
Comment (mtrmac) [inline notary/notary.spec]: But `%check` does run the tests, so this `BuildRequires` seems appropriate.

(Well, maintaining all of the `!with_bundled` case in the spec file is dubious, but I think it is still useful to keep this flexibility to flip this on in the future.)

File: notary/notary.spec (hunk 2 above)
Comment (mtrmac) [inline notary/notary.spec]: `%check` does run the tests.

File: notary/notary.spec (hunk 3 above)
Comment (mtrmac) [inline notary/notary.spec]: `%check` does run the tests.

File: notary/notary.spec (hunk 4 above)
Comment (mtrmac) [inline notary/notary.spec]: Done, thanks!

File: notary/notary.spec (hunk 5 above)
Comment (mtrmac) [inline notary/notary.spec]: Done.

File: notary/notary.spec (hunk 6 above)
Comment (mtrmac) [inline notary/notary.spec]: Done.

File: notary/notary.spec (hunk 7 above)
Comment (mtrmac) [inline notary/notary.spec]: Dropped.

File: notary/notary.spec (hunk 8 above)
Comment (mtrmac) [inline notary/notary.spec:199]: I have kept the `BuildRequires:` in a single block at the top level instead. (They are not specific to subpackages.)

Added the `Requires:`.

(FWIW I am not certain why packaging the unit tests is supposed to be useful. They are not being developed as an API to be imported by anyone else.)

File: notary/notary.spec (hunk 9 above)
Comment (mtrmac) [inline notary/notary.spec]: Good point, dropped.

**Comment by mtrmac** [comment]
Sorry, looking at the full picture, I get the “`golang(github.com/docker/distribution/registry/auth/silly)` can be removed, it is required by tests only” comments now: `%check` is conditional on `with_unit_test`, so the `BuildRequires` can be moved to that block.

OTOH I still can't see why `gofed` is generating a `%package unit_test`, and AFAICS the main block of `BuildRequires` whould not be conditional on `with_devel`

File: notary/notary.spec (hunk 1 above)
Comment (mtrmac) [inline notary/notary.spec]: Never mind, moved to the

```
%if 0%{?with_unit_test} && 0%{?with_devel}
%package unit-test-devel
%if 0%{?with_check}
```

section.

File: notary/notary.spec (hunk 2 above)
Comment (mtrmac) [inline notary/notary.spec]: Never mind, moved to the

```
%if 0%{?with_unit_test} && 0%{?with_devel}
%package unit-test-devel
%if 0%{?with_check}
```

section.

File: notary/notary.spec (hunk 3 above)
Comment (mtrmac) [inline notary/notary.spec]: Never mind, moved to the

```
%if 0%{?with_unit_test} && 0%{?with_devel}
%package unit-test-devel
%if 0%{?with_check}
```

section.

File: notary/notary.spec (hunk 8 above)
Comment (mtrmac) [inline notary/notary.spec:199]: Never mind, done exactly what you suggested.

File: notary/notary.spec (hunk 1 above)
Comment (ingvagabund) [inline notary/notary.spec]: unit-test is here to be run in CI (which is not ready but it is planned). At the end I would like to periodically run unit-tests of all golang packages.

File: notary/notary.spec (hunk 8 above)
Comment (ingvagabund) [inline notary/notary.spec:199]: unit-test is here to be run in CI (which is not ready but it is planned). At the end I would like to periodically run unit-tests of all golang packages.

**Comment by ingvagabund** [comment]
Not every spec file ships devel subpackage. Unit-test is here to isolate additional dependencies so devel subpackage is lighter. Secondly, so scans can see what additional deps I need to run tests. And last but not least to be able to run CI over all golang packages.

**Comment by ingvagabund** [comment]
Next step is to deal with bundled vs. debundled dependencies. Would you be willing to partial build from bundled and partially from debundled? Atm there are dependencies not yet packaged in Fedora. I would recommend to generate spec file for them too, build from debundled dependencies up to docker which would be picked from bundled dependencies. Atm there is no command which would generate that for you. However, it is just a minor change in the spec file. I can help you with that.

**Comment by mtrmac** [comment]
(Thanks for all your work, I have pushed a version fixing those mistakes.)

> Would you be willing to partial build from bundled and partially from debundled? Atm there are dependencies not yet packaged in Fedora. I would recommend to generate spec file for them too, 

I’d be happy to debundle and take advantage of others’ expertise for the various dependencies; I’d rather not own and maintain those dependencies myself, though.

> build from debundled dependencies up to docker which would be picked from bundled dependencies

Looking at the bundled. vs in-distro (F22) versions, by commit time only:

## Older in Fedora

| Package | Status |
| --- | --- |
| github.com/Sirupsen/logrus* | 0.7.3 vs. 0.6.2 in distro, 0.8.4. in updates-testing (?!) |
| github.com/prometheus/client_golang/prometheus | 0.7.0-53-g449ccef vs. 0.5. in distro |
| github.com/golang/protobuf/proto | 3d2510a4dd961caffa2ae781669c628d82db700a (Oct 2 2015) vs. efd7476 (Apr 30 2015) in distro |
| github.com/spf13/cobra | 8e91712f174ced10270cf66615e0a9127e7c4de5 (Jan 17 2016) vs. 0.9.git8f5946c (May 21 2015) in distro |
| golang.org/x/net/context | 47990a1ba55743e6ef1affd3a14e5bac8553615d (Oct 06 2015) vs. 0.25.git446d52d (Jul 21 2015) in distro |
| google.golang.org/grpc* | 3e7b7e58f491074e9577050058fb95d2351a60b0 (Oct 3 2015) vs. 0.1.gitd286668 (Jul 23 2015) in distro |

## Same in Fedora

| Package | Status |
| --- | --- |
| github.com/google/gofuzz | bbcb9da2d746f8bdbd6a936686a0a6067ada0ec5 same as in distro |
| github.com/mattn/go-sqlite3 | 1.0.0 same as in distro |
| github.com/spf13/viper | be5ff3e4840cf692388bde7a057595a474ef379e same as in distro |
| github.com/stretchr/testify* | v1.0-17-g089c718 same as 1.0-0.3.git089c718 in distro |

## Newer in Fedora

| Package | Status |
| --- | --- |
| github.com/docker/docker/pkg/term | 1.7.1 vs. 1.8.2 in distro |
| github.com/gorilla/mux | e444e69cbd2e2e3e0749a2f3c717cec491552bbf (Sep 26 2014) vs. 8096f47 (Mar 11 2015) in distro |
| golang.org/x/crypto* | bfc286917c5fcb7420d7e3092b50bbfd31b38a98 (Jan 24 2015) vs. 0.3.gitc57d4a7 (Mar 27 2015) in distro |

## Not packaged in Fedora

| Package | Status |
| --- | --- |
| github.com/agl/ed25519 | not packaged |
| github.com/bugsnag/bugsnag-go | not packaged |
| github.com/docker/distribution* | not packaged |
| github.com/docker/go-connections/tlsconfig | not packaged |
| github.com/docker/go/canonical/json | not packaged |
| github.com/dvsekhvalnov/jose2go | not packaged |
| github.com/go-sql-driver/mysql | not packaged |
| github.com/jinzhu/gorm | not packaged |
| github.com/miekg/pkcs11 | not packaged |

---

I didn’t try to compare the APIs or do any real analysis of the substantial differences; are there any tools for that?

**Comment by mtrmac** [comment]
Trying further,
- golang-github-spf13-viper-devel-0-0.1.gitbe5ff3e.fc22 is available but has missing dep golang(github.com/spf13/cast)
- google.golang.org/grpc\* API has changed, using the old version breaks the build
- github.com/spf13/cobra API has changed, using the old version breaks the build

Apart of these three issues, unbundling all dependencies which are already packaged in Fedora does allow the package to build and pass its tests. (Of course, I have no idea whether it fully works, or whether it is perhaps broken or insecure due to using an old or unexpected version of a dependency.)

**Comment by ingvagabund** [comment]
With this [1] I am able to build notary locally. There are 6 packages missing in Fedora [2]. Plus docker distribution has to be update to provide devel subpackage (already fix locally, waiting for 3 packages from [2]). Once all new spec [2] are reviewed and built in Fedora, notary can be built from debundled deps.

[1] https://github.com/gofed/reviews/commit/a4050519e547eadd3d54dac9b571f7ef420c45a2
[2] https://bugzilla.redhat.com/buglist.cgi?quicksearch=1314977%201314978%201314979%201315099%201313942%2C%201314213&list_id=4739512

**Comment by ingvagabund** [comment]
All outdated deps are updated as well

**Comment by mtrmac** [comment]
Nit: That commit ends up with many `BuildRequires` duplicated in the `%if ! 0%{?with_bundled}` case and in the `%if 0%{?with_devel}`→`%if 0%{?with_check} && ! 0%{?with_bundled}` case. The latter condition is implied by the former condition, so most of the `BuildRequires` only need to be present in the first list.

**Comment by ingvagabund** [comment]
Yeah, one day you can decide not to provide devel subpackage and built from debundled. This is a temporary ugly redundancy. Which I hope get removed with better tooling in future. Another temporary solution is to delete the first `%if ! 0%{?with_bundled}` wrapper and change `%if ! 0%{?with_devel}` to `%if ! 0%{?with_devel} || ! 0%{?with_bundled}` so everytime you built form debundled deps you must build devel as well. The latter seems better.

**Comment by ingvagabund** [comment]
Still, all golang project spec file are subject to change and the spec file structure is still evolving as new associations are discovered.

**Comment by mtrmac** [comment]
> Yeah, one day you can decide not to provide devel subpackage and built from debundled. This is a temporary ugly redundancy. 

I can’t see that; aren’t `BuildRequires` per-SRPM? AFAIK it cannot matter whether the `BuildRequires` is in the top-level package or in the `%package devel` section; all of them must be present to build from the SRPM. So, if there is a `BuildRequires` with only the `!with_bundled` condition, the same `BuildRequires` with a stricter condition, regardless of which `%package` it is in, should be redundant.

**Comment by mtrmac** [comment]
> I can’t see that; aren’t `BuildRequires` per-SRPM? AFAIK it cannot matter whether the `BuildRequires` is in the top-level package or in the `%package devel` section

Yes, looking at `rpm -qRp $notary.src.rpm` this really does not matter.

I have added more specific comments on this to https://github.com/gofed/reviews/commit/a4050519e547eadd3d54dac9b571f7ef420c45a2 .

Also, notary no longer(?) builds on F24:

```
No matching package to install: 'golang(github.com/docker/distribution/context)'
No matching package to install: 'golang(github.com/docker/distribution/health)'
No matching package to install: 'golang(github.com/docker/distribution/registry/api/errcode)'
No matching package to install: 'golang(github.com/docker/distribution/registry/auth)'
No matching package to install: 'golang(github.com/docker/distribution/registry/auth/htpasswd)'
No matching package to install: 'golang(github.com/docker/distribution/registry/auth/silly)'
No matching package to install: 'golang(github.com/docker/distribution/registry/auth/token)'
No matching package to install: 'golang(github.com/docker/distribution/registry/client/auth)'
No matching package to install: 'golang(github.com/docker/distribution/registry/client/transport)'
No matching package to install: 'golang(github.com/docker/distribution/uuid)'
```

… and the RHEL build is failing with the https://github.com/gofed/reviews/commit/a4050519e547eadd3d54dac9b571f7ef420c45a2 commit as well; at least that one is easy to fix.

**Comment by ingvagabund** [comment]
Here is the refinement of the main package section and the %patch1 wrapped [1]

[1] https://github.com/gofed/reviews/commit/8691966b14029b0a7586a20b7b3ea5f43fa672b4

**Comment by ingvagabund** [comment]
> Also, notary no longer(?) builds on F24:

That is expected as there is no devel subpackage for f24. We can fix it later.

**Comment by mtrmac** [comment]
I'm afraid I don't plan to work on this package any more; the Fedora bug at https://bugzilla.redhat.com/show_bug.cgi?id=1310886 has been closed.

**Pull request closed without merging**